### PR TITLE
[CodeCompletion] Re-implement generic requirement completion

### PR DESCRIPTION
**Pull request by rintaro**
- Rename `completeGenericParam()` -> `completeGenericRequirement()` because it also suggests `associatedtype`s
- Get the target generic context from the current decl context
- Generalize `completeGenericRequirement()` to handle all `where` clause (not only `extension` decl).
- Suggest `associatedtype`s in protocol context rdar://problem/58580482 
- Suggest `typealias`es as long as they are aliases of associated types, generic parameters, or their dependent member types.

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test Linux

**Comment by rintaro** [comment]
@swift-ci Please smoke test Linux

File: lib/IDE/CodeCompletion.cpp [hunk 1]
```diff
@@ -1565,6 +1566,34 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
...
+      return false;
+    archeTy = archeTy->getRoot();
+
+    // For protocol, the 'archeTy' should match with the 'baseTy' which is the
+    // dynamic 'Self' type of the protocol.
+    if (archeTy->isEqual(selfTy))
+      return true;
+
+    // For nominal decls, 'archTy' should be one of the generic parameters.
+    if (selfTy->castTo<BoundGenericType>() &&
```
Comment (benlangmuir) [inline lib/IDE/CodeCompletion.cpp]: `castTo` is not allowed to return nullptr; did you mean to use `getAs`?

Comment (rintaro) [reply lib/IDE/CodeCompletion.cpp]: You're right. I meant `getAs`. But I reconsider it, and updated to find `archeTy` from `selfTy`.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by benlangmuir — APPROVED** [review]

**Comment by rintaro** [comment]
@swift-ci Please smoke test Linux

**Pull request merged**